### PR TITLE
Update workitem validation logic

### PR DESCRIPTION
**Pull request by aholstrup1**
#### Summary 

Updated the workitem validation such that
* All PRs from forks must reference a GitHub issue 
* All PRs must reference an ADO workitem (regardless of whether they're from a fork or not) 

To help link BCIdea workitems to pull requests, I've added a stage to the WorkitemValidation.yaml that looks for ADO workitems in the linked issues and automatically links them to the pull request.

#### Work Item(s) 
Fixes AB#476348

### CODE REVIEW

File: build/scripts/GitHub/GitHubPullRequest.class.psm1
Comment (mazhelez) [inline build/scripts/GitHub/GitHubPullRequest.class.psm1]: Is the name so generic, so that in the future there might be more fields to update? In any case, some initial doc is a good idea.

File: build/scripts/GitHub/GitHubIssue.class.psm1
Comment (mazhelez) [inline build/scripts/GitHub/GitHubIssue.class.psm1:32]: This check is missing the PR class. I wonder if we missed it or it was intentional. 😄

File: build/scripts/GitHub/GitHubIssue.class.psm1
Comment (mazhelez) [inline build/scripts/GitHub/GitHubIssue.class.psm1]: The fact that this was copy-pasted (based on the doc above) shows that this functionality is better suited in a common class. What do you think?

Comment (mazhelez) [inline build/scripts/GitHub/GitHubIssue.class.psm1]:
```suggestion
    [int[]] GetLinkedADOWorkItems() {

```

File: build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1
Comment (mazhelez) [inline build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1]:
```suggestion
                Write-Host "Pull request already linked to ADO workitem AB#$($adoWorkItem)"

```

`$adoWorkItem` is an _int_ right?

File: build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1
Comment (mazhelez) [inline build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1]:
```suggestion
        Write-Host "Trying to link work items from $issueId to pull request $($PullRequest.PRNumber)"

```
Maybe?

File: build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1
Comment (mazhelez) [inline build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1]: This parameter shouldn't be needed. Or at least I don't see a case when it's not `$PullRequest.Repository`.

File: build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1
Comment (mazhelez) [inline build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1]: More like...

```suggestion
    # Find all ADO work items linked to the provided issues and link them to the PR
```

File: build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1
Comment (mazhelez) [inline build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1]: Similar to `GetLinkWorkItemIDs` in the PR class, can we have a `IsLinkedToWorkItem($workItemID)`?

Or `LinkToWorkItem($workItemID)` that links the PR to the work item (if not already linked). E.g. adds "Fixes AB#$workItemID" to the PR description?

My idea is to have the "linking logic" (the AB# syntax) hidden in the classes.

File: build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1
Comment (mazhelez) [inline build/scripts/PullRequestValidation/LinkPullRequestToWorkItem.ps1]: Defensive check might be needed here, `$issue` might be `$null`.

File: build/scripts/PullRequestValidation/ValidateInternalWorkitemForPullRequest.ps1
Comment (mazhelez) [inline build/scripts/PullRequestValidation/ValidateInternalWorkitemForPullRequest.ps1]: This is fine for MS devs, but for non-MS devs might be better to show something like "PRs cannot be merged if not linked to an internal work item. Repo maintainers are responsible for linking PRs. Please do not add any links to internal work items on your own". Too nice? 😄

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [inline .github/workflows/WorkitemValidation.yaml]: Why a separate job for the milestone? Could just be a last step in the previous job. Or am I missing your point?

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [inline .github/workflows/WorkitemValidation.yaml]:
```suggestion
  GitHubIssueValidation:
```

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [inline .github/workflows/WorkitemValidation.yaml]: Why not get the two jobs run in parallel? These are two different and independent checks.

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [inline .github/workflows/WorkitemValidation.yaml]: This step seems better suited for _WorkitemValidationForMicrosoft_ job. 

If it manages to link the PR to a work item, then everything will be nice and green.

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [inline .github/workflows/WorkitemValidation.yaml]:
```suggestion
  WorkItemValidationForMicrosoft:
```

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [inline .github/workflows/WorkitemValidation.yaml]:
```suggestion
    name: 'For Microsoft: Validate link to internal work items' 
```

File: build/scripts/PullRequestValidation/ValidateInternalWorkitemForPullRequest.ps1
Comment (mazhelez) [reply build/scripts/PullRequestValidation/ValidateInternalWorkitemForPullRequest.ps1]: Now I see the job is called "For Microsoft: Validate link to internal work items", so if you think that's enough, I'm also fine with it.

**Review by mazhelez — COMMENTED** [review]
This effort is shaping quite nicely! Well done!
I added a few suggestions for improvements. Let me know what you think.

File: .github/workflows/WorkitemValidation.yaml
Comment (aholstrup1) [reply .github/workflows/WorkitemValidation.yaml]: We can.. The idea behind splitting them was that the previous job will most likely become a required check. I didn't think we should block PRs from merging if linking the PR to a milestone failed for some reason.

File: .github/workflows/WorkitemValidation.yaml
Comment (aholstrup1) [reply .github/workflows/WorkitemValidation.yaml]: Mmh.. I see your point. However, WorkItemValidationForMicrosoft will try to find AB# references in the linked issue. Wouldn't it be nice to know that the issue is validated (i.e. open and approved) before we do this?

File: build/scripts/GitHub/GitHubPullRequest.class.psm1
Comment (aholstrup1) [reply build/scripts/GitHub/GitHubPullRequest.class.psm1]: That was my idea, yes :) I'll add some docs

Comment (aholstrup1) [reply build/scripts/GitHub/GitHubPullRequest.class.psm1]: Actually, for now let's make it UpdateDescription. Then we can make an update function later if needed

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [reply .github/workflows/WorkitemValidation.yaml]: Yeah, that's good point, let's keep the chain.

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [reply .github/workflows/WorkitemValidation.yaml]: If we are to trust milestones, then PRs needs to be linked to milestones always. 

The logic isn't that complex, so I don't expect to have it failing regularly.

File: build/scripts/GitHub/GitHubIssue.class.psm1
Comment (aholstrup1) [reply build/scripts/GitHub/GitHubIssue.class.psm1]: I created a new helper class for this purpose. Let me know what you think

File: build/scripts/GitHub/GitHubIssue.class.psm1
Comment (aholstrup1) [reply build/scripts/GitHub/GitHubIssue.class.psm1:32]: I think we should. Added it now + added some defensive checks in case the PR cannot be found for some reason

File: build/scripts/GitHub/GitHubWorkitemLink.class.psm1
Comment (mazhelez) [inline build/scripts/GitHub/GitHubWorkitemLink.class.psm1]:
```suggestion
class GitHubWorkItemLink {

```

Comment (mazhelez) [reply build/scripts/GitHub/GitHubWorkitemLink.class.psm1]: Also, file name.

File: build/scripts/GitHub/GitHubWorkitemLink.class.psm1
Comment (mazhelez) [inline build/scripts/GitHub/GitHubWorkitemLink.class.psm1]:
```suggestion
    static [int[]] GetLinkedADOWorkItemIDs($Description) {

```

to match `GetLinkedIssueIDs`

File: build/scripts/GitHub/GitHubWorkitemLink.class.psm1
Comment (mazhelez) [inline build/scripts/GitHub/GitHubWorkitemLink.class.psm1]: Why is "Pull Request" inferred? 
Maybe `LinkToADOWorkItem` can return _true_/_false_ depending if it updated `$Description`, which can be passed as _ref_?

File: .github/workflows/WorkitemValidation.yaml
Comment (mazhelez) [inline .github/workflows/WorkitemValidation.yaml]:
```suggestion
          build/scripts/PullRequestValidation/ValidateInternalWorkItemForPullRequest.ps1 -PullRequestNumber ${{ github.event.pull_request.number }} -Repository ${{ github.repository }}
```

**Review by mazhelez — DISMISSED** [review]
Minor tweaks are needed, but over all LGTM.

**Review by ankosiorek — APPROVED** [review]

**Pull request merged**
